### PR TITLE
`dynamic_one_shot` uses tapes with shot-vectors and jitting takes advantage of it

### PR DESCRIPTION
**Pull request by vincentmr**
### Before submitting

Please complete the following checklist when submitting a PR:

- [x] All new features must include a unit test.
      If you've fixed a bug or added code that should be tested, add a test to the
      test directory!

- [x] All new functions and code must be clearly commented and documented.
      If you do make documentation changes, make sure that the docs build and
      render correctly by running `make docs`.

- [ ] Ensure that the test suite passes, by running `make test`.

- [x] Add a new entry to the `doc/releases/changelog-dev.md` file, summarizing the
      change, and including a link back to the PR.

- [x] The PennyLane source code conforms to
      [PEP8 standards](https://www.python.org/dev/peps/pep-0008/).
      We check all of our code against [Pylint](https://www.pylint.org/).
      To lint modified files, simply `pip install pylint`, and then
      run `pylint pennylane/path/to/file.py`.

When all the above are checked, delete everything above the dashed
line and fill in the pull request template.

------------------------------------------------------------------------------------------------------------

**Context:**
`dynamic_one_shot` creates n-shots tapes which is wasteful.

**Description of the Change:**
Create a single tape with a shot-vector which indicates to the device how many times to repeat the tape execution.

**Benefits:**
For a tape like
```
dev = qml.device("default.qubit", shots=2000, seed=jax.random.PRNGKey(123))

@qml.qnode(dev, diff_method=None)
def func(x, y):
    qml.RX(x, wires=0)
    m0 = qml.measure(0, reset=False, postselect=1)
    qml.cond(m0, qml.RY)(y, wires=1)
    return qml.expval(qml.PauliZ(0))


params = np.pi / 4 * np.ones(2)
```
The execution times are as follows (Latitude laptop):
- 12.7 s : vanilla Python
- 8.2 s : jax.vmap
- 11.1 s : jax.jit + jax.vmap + compilation
- 6.89 ms : jax.jit + jax.vmap

**Possible Drawbacks:**

**Related GitHub Issues:**
[sc-62097]

### CODE REVIEW

File: pennylane/devices/default_qubit.py [hunk 1]
```diff
@@ -591,7 +591,7 @@ def execute(
         prng_keys = [self.get_prng_keys()[0] for _ in range(len(circuits))]
 
         if max_workers is None:
-            return tuple(
+            results = tuple(
```
Comment (albi3ro) [inline pennylane/devices/default_qubit.py:594]: ?

File: pennylane/devices/qubit/simulate.py [hunk 2]
```diff
@@ -279,9 +279,33 @@ def simulate(
...
+        if qml.math.get_deep_interface(circuit.data) == "jax":
+            # pylint: disable=import-outside-toplevel
+            import jax
+
+            def vv(k):
+                return simulate_one_shot_native_mcm(
+                    aux_circ, debugger=debugger, rng=rng, prng_key=k, interface=interface
+                )
+
+            results = jax.vmap(vv, in_axes=(0,))(keys)
```
Comment (dime10) [inline pennylane/devices/qubit/simulate.py:299]: vmap, nice 👍 Looks like the randomness is also handled correctly :) Do you know if this speeds up both jitted *and* non-jitted execution?

**Review by dime10 — APPROVED** [review]
This looks good to me 👍

File: pennylane/_qubit_device.py [hunk 3]
```diff
@@ -275,8 +275,19 @@ def execute(self, circuit, **kwargs):
         self.check_validity(circuit.operations, circuit.observables)
 
         has_mcm = any(isinstance(op, MidMeasureMP) for op in circuit.operations)
-        if has_mcm:
-            kwargs["mid_measurements"] = {}
+        if has_mcm and "mid_measurements" not in kwargs:
```
[the remark this answers is not in the document]
Comment (vincentmr) [reply pennylane/_qubit_device.py:278]: It is used as a trigger to call `self.execute` recursively.

File: pennylane/devices/qubit/simulate.py (hunk 2 above)
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:299]: Answered in the description.

**Review by albi3ro — COMMENTED** [review]
Changelog?

File: pennylane/devices/qubit/simulate.py [hunk 4]
```diff
@@ -279,9 +279,33 @@ def simulate(
...
+            circuit.measurements,
+            shots=[1],
+            trainable_params=circuit.trainable_params,
         )
+        keys = jax_random_split(prng_key, num=circuit.shots.total_shots)
+        if qml.math.get_deep_interface(circuit.data) == "jax":
+            # pylint: disable=import-outside-toplevel
+            import jax
+
+            def vv(k):
```
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:294]: vv? What's that stand for?

File: pennylane/devices/qubit/simulate.py [hunk 5]
```diff
@@ -279,9 +279,33 @@ def simulate(
...
+                return simulate_one_shot_native_mcm(
+                    aux_circ, debugger=debugger, rng=rng, prng_key=k, interface=interface
+                )
+
+            results = jax.vmap(vv, in_axes=(0,))(keys)
+            results = list(zip(*results))
```
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:300]: Any particular reason to use `list` over `tuple`?

File: pennylane/devices/qubit/simulate.py [hunk 6]
```diff
@@ -279,9 +279,33 @@ def simulate(
...
+        else:
+            for i, _ in enumerate(circuit.shots):
```
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:302]:
```suggestion
            for i in range(circuit.shots.total_shots):
```

purely stylistic.

File: pennylane/devices/qubit/simulate.py [hunk 7]
```diff
@@ -279,9 +279,33 @@ def simulate(
...
+                results.append(
+                    simulate_one_shot_native_mcm(
+                        aux_circ, debugger=debugger, rng=rng, prng_key=keys[i], interface=interface
```
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:305]: If we already know there's no jax involved at this point, do we even need the `prng_key`? Isn't that basically only for jax-native simulations?

File: pennylane/devices/qubit/simulate.py (hunk 2 above)
Comment (albi3ro) [inline pennylane/devices/qubit/simulate.py:299]: What would happen if a user is performing `vmap` higher in the stack?  Do nested vmap's work?

File: pennylane/devices/qubit/simulate.py (hunk 4 above)
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:294]: Nothing really, it is just a local function definition for vmap. Would you prefer we call it `simulate_partial`?

File: pennylane/devices/qubit/simulate.py (hunk 5 above)
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:300]: No, I'll change it.

File: pennylane/devices/qubit/simulate.py (hunk 7 above)
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:305]: Correct, let me change it and see if anything breaks.

File: pennylane/devices/qubit/simulate.py (hunk 2 above)
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:299]: Yes, you can do something like
```
A = np.random.rand(10, 10, 10)
B = np.random.rand(10, 10, 10)
f = jax.vmap(lambda x, y: jnp.vdot(x, y), (0, 1))
C = jax.vmap(lambda x, y: f(x, y), (0, 1))(A, B)
```

File: pennylane/devices/qubit/simulate.py (hunk 7 above)
Comment (vincentmr) [reply pennylane/devices/qubit/simulate.py:305]: `interface` here is a config keyword, not necessarily matching the deep interface of the tape data. We allow using JAX's RNG for sampling with NumPy arrays, in which case we should still pass the RNG keys.

**Review by mudit2812 — APPROVED** [review]

**Review by albi3ro — APPROVED** [review]
Might be worth investigating if we can somehow use use multi-processing with this in the future.

Thanks for doing all this investigation!

**Comment by vincentmr** [comment]
> Might be worth investigating if we can somehow use use multi-processing with this in the future.

That's actually the original idea behind having multiple tapes instead of shots. On the other hand, I think multi-processing might be better turned-on at the QNode level, or maybe with a decorator. I think any device with a batch of tapes could be executing them in parallel inside sub-processes.

**Comment by dime10** [comment]
> > Might be worth investigating if we can somehow use use multi-processing with this in the future.
> 
> That's actually the original idea behind having multiple tapes instead of shots. On the other hand, I think multi-processing might be better turned-on at the QNode level, or maybe with a decorator. I think any device with a batch of tapes could be executing them in parallel inside sub-processes.

Why couldn't you parallelize with shots instead of tapes?

**Comment by vincentmr** [comment]
> Why couldn't you parallelize with shots instead of tapes?

The current multiprocessing logic live [here](https://github.com/PennyLaneAI/pennylane/blob/master/pennylane/devices/default_qubit.py#L608). It is not a fundamental limitation, it's just our UI is not set up for it.
